phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java

```diff
@@ -279,6 +301,31 @@ protected RegionScanner getWrappedScanner(final ObserverContext<RegionCoprocesso
             final ImmutableBytesWritable ptr) {
         return new RegionScanner() {
 
+            private boolean hasReferences = checkForReferenceFiles();
+            private HRegionInfo regionInfo = c.getEnvironment().getRegionInfo();
+            private byte[] actualStartKey = getActualStartKey();
+
+            private boolean checkForReferenceFiles(){
+                for(byte[] family: scan.getFamilies()) {
+                    if(c.getEnvironment().getRegion().getStore(family).hasReferences()) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            public byte[] getActualStartKey() {
+                if(ScanUtil.isLocalIndex(scan) && scan.getAttribute(SCAN_START_ROW_SUFFIX)!=null) {
+                    byte[] startKey = ScanRanges.prefixKey(scan.getAttribute(SCAN_START_ROW_SUFFIX), 0, regionInfo.getStartKey().length == 0 ? new byte[regionInfo.getEndKey().length]: regionInfo.getStartKey(), regionInfo.getStartKey().length == 0? regionInfo.getEndKey().length: regionInfo.getStartKey().length);
+                    if(Bytes.compareTo(scan.getStartRow(), startKey)>=0) {
+                        return scan.getStartRow();
+                    } else {
+                        return startKey;
+                    }
+                }
+                return null;
+            }
+
             @Override
             public boolean next(List<Cell> results) throws IOException {
```

Comment:
Can we refactor the code between these two next(List<Cell>, ...) implementations, as there's now a fair amount of code which is duplicated here?